From 4d1b9bcef1d9d82a16cf4b1a50271bad8933990f Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Tue, 15 Nov 2022 09:24:49 +0200
Subject: [PATCH] [SW-0] gaudi3: init psoc arcs cpu id

In order for the FW to be able to get the CPU ID on which the code is
executing, this ID must be written before the CPU reset is released.
Since ARC1 is the first to run (by HW), its ID will be the default (0).
For the other CPUs a relative ID is set. So the code can easily
normalize the cpu ID by adding '1' to whatever is read from the
IDENTITY register.

Change-Id: I94882c11adc14b0188295e5af2b975d74d5d6a84
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 5585f43..3aaeac2 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -17,6 +17,8 @@
 #include <d0_psoc_uart_0_regs.h>
 #include <d0_psoc_timestamp_regs.h>
 #include <d0_parc_arc0_cfg_regs.h>
+#include <d0_parc_arc1_cfg_regs.h>
+#include <d0_parc_arc2_cfg_regs.h>
 #include <d0_psoc_i2c_s_regs.h>
 #include <d0_parc_int_aggr_i2c_slv_comb_regs.h>
 #include <d0_parc_local_cfg_regs.h>
@@ -163,11 +165,29 @@
 
 __weak void hl_dev_cpu_pll_init(void) {}
 
+/*
+ * CPU ID must be written before the CPU reset is released. Since ARC1 is the first
+ * to run (by HW), its ID will be the default (0). For the other CPUs a relative ID
+ * is set. So the code can easily normalize the cpu ID by adding '1' to whatever is
+ * read from the IDENTITY register.
+ */
+#define PSOC_ARC0_PRELOADED_ID	-1
+#define PSOC_ARC1_PRELOADED_ID	0
+#define PSOC_ARC2_PRELOADED_ID	1
+
+static void hl_init_cpu_ids(void)
+{
+	regmap_writel(PSOC_ARC0_PRELOADED_ID, mmD0_PARC_ARC0_CFG_ARC_NUM);
+	regmap_writel(PSOC_ARC1_PRELOADED_ID, mmD0_PARC_ARC1_CFG_ARC_NUM);
+	regmap_writel(PSOC_ARC2_PRELOADED_ID, mmD0_PARC_ARC2_CFG_ARC_NUM);
+}
+
 void hl_plat_early_init(void)
 {
 	hl_lbw_early_write_resp_enable();
 	init_prstn();
 	hl_dev_cpu_pll_init();
+	hl_init_cpu_ids();
 }
 __weak void hl_dev_nw_pll_init(void) {}
 
